### DISCUSSION TITLE
support `vcpkg export` in manifest mode

### DISCUSSION THREAD

**Discussion by Be-ing** [microsoft/vcpkg#17873] [Ideas]
**Is your feature request related to a problem? Please describe.**
Manifest mode introduced a handful of nice features like allowing platform-specific definitions of dependencies. I would like to use this together with `vcpkg export`. Currently, attempting to do this fails with:
```
vcpkg export does not support manifest mode, in order to allow for future design considerations. You may use export in classic mode by running vcpkg outside of a manifest-based project.
Note: Updating vcpkg by rerunning bootstrap-vcpkg may resolve this failure.
Error: Process completed with exit code 1.
```

**Proposed solution**
`vcpkg export` works as it does now when running without a vcpkg.json file.

**Describe alternatives you've considered**
Currently we define a big list of packages in a GitHub Actions workflow file in our fork of the vcpkg repo. Customizing this for different OSes would require implementing that logic in the GH Actions workflow file. It would be nice to move this out of the CI script and into a separate file, which would allow using the hash of the vcpkg.json file in GH Actions cache keys.

**Comment by PhilippGrulich** [comment]
Hey,

i think that would be a very cool feature!

**Comment by Be-ing** [comment]
What would it take to implement this? Is there any architectural issue in the way?

**Comment by hellozee** [comment]
Sharing the `vcpkg_installed` directory as an archive should just work, unless I am missing something since I haven't tried that yet.

**Reply by eladmaimoni** [reply]
You would have to somehow manually integrate it to the build system (cmake / msbuild). How would you do this?

The export command also creates a cmake script that can be used as a cmake toolchain file.

**Reply by autoantwort** [reply]
Simply copy the files from scripts/buildsystems

**Reply by hyt589** [reply]
I just tried this and vcpkg just deleted everything from the shared `vcpkg_installed` and started from scratch.

**Comment by morphogencc** [comment]
I would love to see this feature as well!  I use manifest mode for almost all of my projects; it provides a convenient way to share dependencies using vcpkg.  However, sometimes I want to deploy the built binary from my project and there's currently no easy way to share all of the integrated dependencies that vcpkg builds for me.

The current error I'm given is:

```
vcpkg export does not support manifest mode, in order to allow for future design considerations. You may use export in classic mode by running vcpkg outside of a manifest-based project.
note: updating vcpkg by rerunning bootstrap-vcpkg may resolve this failure.
```

However, it seems that it's been this way for more than 18 months and there's no clear indication what the roadmap is, or what technical issues are preventing this from working.

**Comment by AugP** [comment]
We recognize that there are a number of commands that do not work with manifest mode today. We intend to rectify this in the future. The same goes for commands for vcpkg artifacts vs. ports. I don't have a date currently for when "vcpkg export" will be looked at, but I will discuss it further with the other maintainers.

**Comment by RealLitb** [comment]
Is there any possibility to work around the limitation? How does VCPKG even know that it is within a manifest-mode installation? I moved vcpkg.json out of the way, but VCPKG still "knows" that the install-root is from a manifest install. 

EDIT: I moved the vcpkg_installed into the VCPKG worktree. Then it appears to work and will export all installed packages. 

What would be cool would be  a command to print out a `vcpkg install` command line parameter list for classic-mode, given a manifest `vcpkg.json` file! Like a `--only-show` flag for `vcpkg install` that  prints the packages that would be installed in manifest mode. Then we could use that output for a classic run of VCPKG which would officially support export.

**Comment by KUGA2** [comment]
Cant we just zip an export the vcpkg_installed folder? Whats wrong with that?

**Reply by dg0yt** [reply]
For CMake, you will still need vcpkg's `vcpkg.cmake` and the chainloaded toolchain files. `vcpkg.cmake` calls the installed wrappers for CMake packages.
Cf. https://github.com/microsoft/vcpkg/discussions/17873#discussioncomment-1359399

**Reply by itfanr** [reply]
I copied the exported folder to a project, and failed to build my project (no CMakeList.txt)

**Comment by AugP** [comment]
Update: we shipped vcpkg export for manifest mode in the 2023.10.19 release: https://devblogs.microsoft.com/cppblog/vcpkg-2023-10-19-release-export-for-manifests-documentation-improvements-and-more/

Let us know what you think!

**Reply by itfanr** [reply]
But how could I import vcpkg pacakges on another PC and use the packages without `vcpkg install`? 

 I don't consider cmake

**Comment by ras0219-msft** [comment]
Update: we shipped vcpkg export for manifest mode in the 2023.10.19 release: https://devblogs.microsoft.com/cppblog/vcpkg-2023-10-19-release-export-for-manifests-documentation-improvements-and-more/
